### PR TITLE
Portal resources

### PR DESCRIPTION
**Pull request by he-is-james**
- implemented adding new, viewing all, editing, and deleting resources
- updated resource model for the information
- adjusted rendering for days of operation

### CODE REVIEW

File: website/src/Pages/ResourcesPage/AddResourcePage.jsx [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            <TimePicker
+              value={endTime}
+              onChange={(newTime) => setEndTime(newTime)}
+            />
+          </LocalizationProvider>
+        </Box>
+      </Box>
+      <Box>
+        <Box>
+          <Typography>Latitude</Typography>
```
Comment (siddharthnandy) [inline website/src/Pages/ResourcesPage/AddResourcePage.jsx:200]: So we separate Latitude, Longitude, and Address fields, which may lead to collisions if inputs don't line up for some reason. I think it would be a good idea to obscure the coordinates stuff from the front-end, and [Google seems to have an API for it](https://developers.google.com/maps/documentation/geocoding/overview). Completely up to you, as this would entail some changes for the "reverse geo-coding" (i.e., deriving the coordinates from the address), which we would need to populate the latitude and longitude coordinates in our database (we would naturally need to perform some validation to make sure 1) the address is valid and 2) it is indeed the correct address the admins want; this could be done perhaps with a confirmation page that shows a map)

TLDR I think having Lat/Long/Addr fields is, at best, inconvenient for the admins and, at worst, a bit risky if admins make a mistake with the inputs

File: website/src/Pages/ResourcesPage/AddResourcePage.jsx [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+      </Box>
+      <Box>
+        <Typography>Address</Typography>
+        <TextField
+          value={address}
+          onChange={(e) => setAddress(e.target.value)}
+        />
+      </Box>
+      <Box>
+        <Typography>Email</Typography>
```
Comment (siddharthnandy) [inline website/src/Pages/ResourcesPage/AddResourcePage.jsx:222]: light regex validation on this field might be a good idea

File: website/src/Pages/ResourcesPage/AddResourcePage.jsx [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        <TextField
+          value={email}
+          onChange={(e) => setEmail(e.target.value)}
+        />
+      </Box>
+      <Box>
+        <Typography>Website</Typography>
```
Comment (siddharthnandy) [inline website/src/Pages/ResourcesPage/AddResourcePage.jsx:229]: light regex validation on this field might be a good idea

File: website/src/Pages/ResourcesPage/AddResourcePage.jsx [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    // Format the phone number once it reaches 10 digits
+    if (numericValue.length === 10) {
+      const formattedValue = `(${numericValue.slice(0, 3)}) ${numericValue.slice(3, 6)}-${numericValue.slice(6, 10)}`;
+      setPhoneNumber(formattedValue);
+    } else if (numericValue.length < 10) {
+      setPhoneNumber(numericValue);
+    }
+  };
+
+  const createResource = async () => {
```
Comment (siddharthnandy) [inline website/src/Pages/ResourcesPage/AddResourcePage.jsx:55]: add validation for fields, if something fails maybe we could display a message of what failed

File: website/src/Pages/ResourcesPage/EditResourcePage.jsx [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    // Format the phone number once it reaches 10 digits
+    if (numericValue.length === 10) {
+      const formattedValue = `(${numericValue.slice(0, 3)}) ${numericValue.slice(3, 6)}-${numericValue.slice(6, 10)}`;
+      setPhoneNumber(formattedValue);
+    } else if (numericValue.length < 10) {
+      setPhoneNumber(numericValue);
+    }
+  };
+
+  const editResource = async () => {
```
Comment (siddharthnandy) [inline website/src/Pages/ResourcesPage/EditResourcePage.jsx:79]: validation (like in `AddResourcePage`)

File: website/src/Pages/ResourcesPage/AddResourcePage.jsx [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+            <MenuItem value="health">health</MenuItem>
+            <MenuItem value="legal services">legal services</MenuItem>
+            <MenuItem value="shower">shower</MenuItem>
+            <MenuItem value="mission">mission</MenuItem>
+            <MenuItem value="social services">social services</MenuItem>
+          </Select>
+        </FormControl>
+      </Box>
+      <Box sx={{ display: 'flex' }}>
+        <Button
```
Comment (siddharthnandy) [inline website/src/Pages/ResourcesPage/AddResourcePage.jsx:121]: there's probably a way to implement this with a map, (use array `[M, TU,W,TH,F,SA,SU]` and use both the elements and indices in the map), but it's up to you/whoever implements styling

File: website/src/Pages/ResourcesPage/EditResourcePage.jsx [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+            <MenuItem value="health">health</MenuItem>
+            <MenuItem value="legal services">legal services</MenuItem>
+            <MenuItem value="shower">shower</MenuItem>
+            <MenuItem value="mission">mission</MenuItem>
+            <MenuItem value="social services">social services</MenuItem>
+          </Select>
+        </FormControl>
+      </Box>
+      <Box sx={{ display: 'flex' }}>
+        <Button
```
Comment (siddharthnandy) [inline website/src/Pages/ResourcesPage/EditResourcePage.jsx:145]: [same as `AddResourcePage`] there's probably a way to implement this with a map, (use array [M, TU,W,TH,F,SA,SU] and use both the elements and indices in the map), but it's up to you/whoever implements styling

**Review by siddharthnandy — CHANGES_REQUESTED** [review]

File: server/models/resourceModel.js [hunk 8]
```diff
@@ -1,5 +1,20 @@
 const mongoose = require('mongoose');
 
+const dateSchema = new mongoose.Schema({
```
[the remark this answers is not in the document]
Comment (he-is-james) [reply server/models/resourceModel.js:3]: I prefer `dateSchema` also a bit of a pain to adjust

File: website/src/Pages/ResourcesPage/AddResourcePage.jsx (hunk 1 above)
Comment (he-is-james) [reply website/src/Pages/ResourcesPage/AddResourcePage.jsx:200]: I agree with the reasoning but because of time restrictions I think it's fine for this process. In the future improvements, it would be better to do all the checking to make the entire admin work easier

File: website/src/Pages/ResourcesPage/AddResourcePage.jsx (hunk 4 above)
Comment (he-is-james) [reply website/src/Pages/ResourcesPage/AddResourcePage.jsx:55]: again, I will leave the alert and validation to the frontend but I implemented preliminary checks @zhangjuliet

File: website/src/Pages/ResourcesPage/AddResourcePage.jsx [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+    if (validateResourceInputs()) {
+      alert('Missing field or incorrect format.');
+      return;
+    }
+
+    const dateDetails = groupDayTimes();
+    const locationDetails = {
+      address,
+      coordinates: {
+        latitude: Number(latitude),
```
Comment (siddharthnandy) [inline website/src/Pages/ResourcesPage/AddResourcePage.jsx:121]: not sure if this is sufficient validation, as when you give this a non-numeric string, latitude becomes NaN and then the server will send a 400 Bad Request reply after making the API call

File: website/src/Pages/ResourcesPage/EditResourcePage.jsx [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+    if (validateResourceInputs()) {
+      alert('Missing field or incorrect format.');
+      return;
+    }
+
+    const dateDetails = groupDayTimes();
+    const locationDetails = {
+      address,
+      coordinates: {
+        latitude: Number(latitude),
```
Comment (siddharthnandy) [inline website/src/Pages/ResourcesPage/EditResourcePage.jsx:149]: same as on AddResourcePage

**Review by siddharthnandy — APPROVED** [review]
one last change but otherwise lgtm
